File: MTLab4JavaWaiter/src/Waiter.java
import java.util.concurrent.Semaphore;

public class Waiter {
    private final Semaphore permission = new Semaphore(4);

    public void requestPermission() {
        try {
            permission.acquire();
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }

    public void releasePermission() {
        permission.release();
    }
}

File: MTLab4Waiter/Waiter.cs
namespace MTLab4Waiter;

public class Waiter
{
	private readonly SemaphoreSlim _permission = new(4, 4);

    public void RequestPermission()
    {
        _permission.Wait();
    }

    public void ReleasePermission()
    {
        _permission.Release();
    }
}